### PR TITLE
Applying same ext_proc filter checker rule for immediate response.

### PR DESCRIPTION
**Pull request by yanjunxiang-google**
Currently, the ext_proc filter Immediate response mutation checker is very permissive. It is allow changing of anything including host/path headers, also can let any headers in without any check.  These may cause security issues.

The solution is to applying same header mutation check rule for immediate response, same as any other header/trailer response.

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by yanavlasov** [comment]
@yanjunxiang-google looks good. Can you add a test, please?

**Comment by yanjunxiang-google** [comment]
> @yanjunxiang-google looks good. Can you add a test, please?

Sure, added two integration tests to just verify immediate response now honors the filter mutation check configuration same as other response messages.  The comprehensive tests for mutation checker are already in place.

**Review by tyxia — COMMENTED** [review]
Instead of removing `ImmediateMutationChecker` which basically disables the original intention that filters can create custom error messages,  we can introduce a new config field (like a flag) and use `config.mutation_checker()` when that field is set to true. By doing this, we can still have original ability when other users of ext_proc want and are guarded against security issues.

**Comment by yanjunxiang-google** [comment]
> Member

This ImmediateMutationChecker https://github.com/envoyproxy/envoy/blob/7c571ac7c61768d72929feda7ef9edeb5046b56f/source/extensions/filters/http/ext_proc/ext_proc.cc#L44
basically has two rules, allow_all_routing(true) and allow_envoy(true).

This basically let the ext_proc server can send back header mutation to add system headers, like :scheme/:authority, or Envoy internal headers "x-envoy-***".  And such headers will be in the local reply messages sent back to client.  This probably is not what Envoy want. Is there a good reason to keep that door open?

**Comment by tyxia** [comment]
> > Member
> 
> This ImmediateMutationChecker
> 
> https://github.com/envoyproxy/envoy/blob/7c571ac7c61768d72929feda7ef9edeb5046b56f/source/extensions/filters/http/ext_proc/ext_proc.cc#L44
> 
> 
> basically has two rules, allow_all_routing(true) and allow_envoy(true).
> This basically let the ext_proc server can send back header mutation to add system headers, like :scheme/:authority, or Envoy internal headers "x-envoy-***". And such headers will be in the local reply messages sent back to client. This probably is not what Envoy want. Is there a good reason to keep that door open?

I think essentially this comes down to one question: Will your PR break other users? From a quick scan, it seems so. For example, what if other users don't have mutation rule set but want it in immediate response OR what if they want to have different mutation rule in immediate response. They could have their own usage we don't know.

But I will let others make the call here.

**Comment by yanavlasov** [comment]
> 

We've decided to allow braking changes for now and transition API to stable after the fix for UTF-8 problem.

File: test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc [hunk 1]
```diff
@@ -1255,6 +1255,44 @@ TEST_P(ExtProcIntegrationTest, GetAndRespondImmediatelyWithBadStatus) {
...
+  proto_config_.mutable_mutation_rules()->mutable_disallow_is_error()->set_value(true);
+  proto_config_.mutable_mutation_rules()->mutable_disallow_system()->set_value(true);
+  initializeConfig();
+  HttpIntegrationTest::initialize();
+  auto response = sendDownstreamRequest(absl::nullopt);
+  processAndRespondImmediately(*grpc_upstreams_[0], true, [](ImmediateResponse& immediate) {
+    immediate.mutable_status()->set_code(envoy::type::v3::StatusCode::Unauthorized);
+    auto* hdr = immediate.mutable_headers()->add_set_headers();
+    // Mutating system header :scheme is not allowed.
+    hdr->mutable_header()->set_key(":scheme");
```
Comment (yanavlasov) [inline test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc:1270]: The `:scheme` header is illegal on the response. Please use the `:status` header.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

**Comment by yanjunxiang-google** [comment]
> 
@tyxia  To satisfy the original goal to have a separated, more liberal set of mutation rules for ImmediateResponse, instead of having it fixed in the Envoy code,  an option is to adding a separate header mutation rule for ImmediateResponse in filter config.

File: test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc (hunk 1 above)
Comment (yanjunxiang-google) [reply test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc]: Sure, it looks like codec has it's own check about what headers can be sent back to client. And it rejected a response with :scheme in it.
Changed to :foo, which can be testing here: https://github.com/envoyproxy/envoy/blob/f00166c4cc95b5d3aea14af40a7afe0e849cc78d/source/extensions/filters/common/mutation_rules/mutation_rules.cc#L82

Comment (yanjunxiang-google) [reply test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc]: BTW, there are existing comprehensive mutation_rule test in test/extensions/filters/common/mutation_rules/mutation_rules_test.cc
google3/third_party/envoy/src/test/extensions/filters/http/ext_proc/mutation_utils_test.cc
test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc
test/extensions/filters/http/ext_proc/filter_test.cc

So, the tests we added are just to verify now the immediate response are also subject to the filter mutation rule checks.

**Comment by yanjunxiang-google** [comment]
We decided to have ext_proc filter to use the same header mutation rule in the filter config to check the header mutations in the immediate response from the ext_proc server.  But added a runtime flag to revert to old behavior.

File: changelogs/current.yaml [hunk 2]
```diff
@@ -15,6 +15,12 @@ behavior_changes:
     When ``append_x_forwarded_host`` is enabled for a given route action it is now only appended iff it is different from the last
     value in the list. This resolves issues where a retry caused the same value to be appended multiple times. This
     behavioral change can be temporarily reverted by setting runtime guard ``envoy_reloadable_features_append_xfh_idempotent`` to false.
+- area: ext_proc
+  change: |
+    ext_proc filter uses hard-coded header mutation rule to handle header mutations in the ImmediateResponse message from the
```
Comment (yanavlasov) [inline changelogs/current.yaml:20]: Suggestion. Replace first 3 sentences to "Apply header mutation rules from the ext_proc config to the ImmediateResponse."

**Review by yanavlasov — COMMENTED** [review]
/wait

**Comment by yanavlasov** [comment]
LGTM with small suggestion for the release note.

File: changelogs/current.yaml (hunk 2 above)
Comment (yanjunxiang-google) [reply changelogs/current.yaml]: Done

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**